pvanalytics/quality/time.py

```diff
@@ -33,3 +36,130 @@ def spacing(times, freq):
+    return np.sign(x) * (quotient*to + remainder)
+
+
+def shifts_ruptures(daytime, clearsky_midday, period_min=2,
+                    shift_min=15, round_up_from=None,
+                    prediction_penalty=13):
+    """Identify time shifts using the ruptures library.
+
+    Compares the solar-transit time from `clearsky_midday` with the
+    mid-point between sunrise and sunset time derived from the `daytime`
+    mask. These times are the times of the first and last True value on
+    each day.
+
+    The Pelt changepoint detection method is applied to the difference
+    in midday times derived from the `daytime` series and the expected
+    midday times from `clearsky_midday`. For each period between change
+    points the mode of the difference is rounded to a multiple of
+    `shift_min` and returned as the time-shift for all values in that
+    period.
+
+    Parameters
+    ----------
+    daytime : Series
+        Boolean time-series with True for day and False for night.
+    clearsky_midday : Series
+        Time of midday in minutes for each day with no time shifts
+        (i.e.based on solar position with a fixed-offset time zone).
+    period_min : int, default 2
+        Minimum number of days between shifts. Must be less than or equal to
+        the number of days in `daytime`.
+
+        Increasing this parameter will make the result less sensitive to
+        transient shifts. For example if your intent is to find and correct
+        daylight savings time shifts passing `period_min=60` can give good
+        results while excluding shorter periods that appear shifted.
+    shift_min : int, default 15
+        Minimum shift amount in minutes. All shifts are rounded to a multiple
+        of `shift_min`. [minutes]
+    round_up_from : int, optional
+        The number of minutes greater than a multiple of `shift_min` for a
+        shift to be rounded up. If a shift is less than `round_up_from` then
+        it will be rounded towards 0. If not specified then the shift will
+        be rounded up from `shift_min // 2`. Using a larger value will
+        effectively make the shift detection more conservative as small
+        variations will tend to be rounded to zero.
+    prediction_penalty : int, default 13
+        Penalty used in assessing change points.
+        See :py:method:`ruptures.detection.Pelt.predict` for more information.
+
+    Returns
+    -------
+    Series
+        Time shift in minutes at each timestamp in `daytime`.
+
+    Raises
+    ------
+    ValueError
+        If the number of days in `daytime` is less than `period_min`.
+
+    Notes
+    -----
+    Derived from the PVFleets QA project.
+
+    """
+    try:
+        import ruptures
+    except ImportError:
+        raise ImportError("time.shifts_ruptures() requires ruptures.")
+
+    midday = _group.by_day(daytime).apply(
+        lambda day: (day[day].index.min()
+                     + ((day[day].index.max() - day[day].index.min()) / 2))
```

Comment:
Not a request to change, but I don't follow `day[day]` here - isn't `day` itself a slice from `daytime`?

Reply:
This is just a simple way to get the first time-stamp that has a `True` value. Since `day` is a Boolean series we can select a slice using itself to get only the daytime timestamps then take the minimum.

Maybe it would be easier to read if we remove the night-time timestamps before the `groupby` operation. Would look something like:

```python
    _group.by_day(daytime[daytime]).apply(lambda day: day.index.min() + (day.index.max() - day.index.min()) / 2)
```